v4/integration/fcm/src/main/java/exchange/dydx/trading/integration/fcm/PushPermissionRequester.kt

```diff
@@ -0,0 +1,94 @@
+package exchange.dydx.trading.integration.fcm
+
+import android.Manifest
+import android.app.Activity
+import android.content.pm.PackageManager
+import android.os.Build
+import androidx.activity.result.ActivityResultCaller
+import androidx.activity.result.ActivityResultLauncher
+import androidx.activity.result.contract.ActivityResultContracts
+import androidx.annotation.RequiresApi
+import androidx.core.app.ActivityCompat.shouldShowRequestPermissionRationale
+import androidx.core.content.ContextCompat
+import dagger.Binds
+import dagger.Module
+import dagger.hilt.InstallIn
+import dagger.hilt.android.components.ActivityRetainedComponent
+import dagger.hilt.android.scopes.ActivityRetainedScoped
+import exchange.dydx.dydxstatemanager.protocolImplementations.AbacusLocalizerImp
+import exchange.dydx.platformui.components.PlatformDialog
+import exchange.dydx.platformui.components.container.PlatformInfo
+import exchange.dydx.utilities.utils.SharedPreferencesStore
+import javax.inject.Inject
+
+@ActivityRetainedScoped
+class RealPushPermissionRequester @Inject constructor(
+    private val platformDialog: PlatformDialog,
+    private val platformInfo: PlatformInfo,
+    private val abacusLocalizerImp: AbacusLocalizerImp,
+    private val sharedPreferencesStore: SharedPreferencesStore,
+) : PushPermissionRequester {
+
+    var requestPermissionLauncher: ActivityResultLauncher<String>? = null
```

Comment:
Where does this get initialized?

Reply:
in the `set()` for `activity` - this API is really awkward and requires an Activity

Reply:
happy to huddle on this to explain what's going on here.

basically you can't get an `Activity` injected via Hilt into a ViewModel, because of the component structure: https://dagger.dev/hilt/components.html

so we need to pass TradingActivity in onCreate/onDestroy to the PushPermissionRequester. when `activity` changes, we need to update `requestPermissionLauncher`.